### PR TITLE
Add user interface builder

### PR DESCRIPTION
**Pull request by Mrtenz**
Work in progress.

### CODE REVIEW

**Comment by Mrtenz** [comment]
@SocketSecurity ignore-all

File: src/features/builder/components/BaseNode.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      paddingY="2"
+      paddingLeft="4"
+      paddingRight="2"
+      background="background.alternative"
+      borderRadius="lg"
+      gap="4"
+      alignItems="center"
+      border="1px solid"
+      borderColor="border.default"
+      // opacity={isDragging ? 0.5 : 1}
```
Comment (FrederikBolding) [inline src/features/builder/components/BaseNode.tsx]:
```suggestion
```

File: src/features/builder/components/NodeRenderer.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * @param props.items - The items to render in the tree.
+ * @returns A node renderer component.
+ */
+export const NodeRenderer: FunctionComponent<NodeRendererProps> = ({
+  items,
+}) => {
+  const node = useMemo(() => nodeModelsToComponent(items), [items]);
+
+  return (
+    <Window snapName="foo-snap" snapId="bar-snap">
```
Comment (FrederikBolding) [inline src/features/builder/components/NodeRenderer.tsx]: Should we show the name of the connected snap here if applicable?

Comment (FrederikBolding) [reply src/features/builder/components/NodeRenderer.tsx]: We can fall back to unknown or something

File: src/features/builder/components/Panel.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { Box } from '@chakra-ui/react';
+import { Component, NodeType, panel } from '@metamask/snaps-ui';
+import { FunctionComponent, useRef, useState } from 'react';
+import { useDrag, useDrop } from 'react-dnd';
+
+export type PanelProps = {
+  // onDrop: (item: Component) => void;
```
Comment (FrederikBolding) [inline src/features/builder/components/Panel.tsx]:
```suggestion
```

File: src/features/builder/Builder.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import { Component, NodeType } from '@metamask/snaps-ui';
+import { NodeModel } from '@minoru/react-dnd-treeview';
+import { FunctionComponent, useState } from 'react';
+
+import { Editor } from '../../components';
+import { TemplateComponentList, NodeTree, NodeRenderer } from './components';
+import { nodeModelsToComponent, panelToCode } from './utils';
+
+export const Builder: FunctionComponent = () => {
+  const [id, setId] = useState<number>(100);
```
Comment (FrederikBolding) [inline src/features/builder/Builder.tsx]: Why are we starting at `100`?

File: src/features/builder/components/Panel.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  nodes?: Component[];
+};
+
+export const Panel: FunctionComponent<PanelProps> = ({ nodes = [] }) => {
```
Comment (FrederikBolding) [inline src/features/builder/components/Panel.tsx]: How is this chosen to be rendered? 🤔

Comment (Mrtenz) [reply src/features/builder/components/Panel.tsx]: It's not. 😅

File: src/features/builder/Builder.tsx (hunk 4 above)
Comment (Mrtenz) [reply src/features/builder/Builder.tsx]: It was done like this in the example. It makes sure there's no overlap between the default nodes and the newly created nodes. In this case we can just start at 2 though.

File: src/features/builder/components/NodeRenderer.tsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+export const NodeRenderer: FunctionComponent<NodeRendererProps> = ({
+  items,
+}) => {
+  const snapName = useSelector(getSnapName) ?? 'Unknown';
+  const node = useMemo(() => nodeModelsToComponent(items), [items]);
+
+  return (
+    <Window snapName={snapName} snapId={DEFAULT_SNAP_ID}>
+      <Box margin="4" marginTop="0" flex="1">
+        <Delineator snapName="foo-snap">
```
Comment (FrederikBolding) [inline src/features/builder/components/NodeRenderer.tsx]:
```suggestion
        <Delineator snapName={snapName}>
```

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**